src/nvim/option.c

```diff
@@ -3518,11 +3517,13 @@ static const char *did_set_option(OptIndex opt_idx, void *varp, OptVal old_value
   }
 
   // Free options that are in allocated memory.
-  // Use "free_oldval", because recursiveness may change the flags (esp. init_highlight()).
-  if (free_oldval) {
+  // NOTE: When initializing an option for the first time, don't try to free its value as it has
+  // not been allocated yet.
+  if (opt_initialized[opt_idx]) {
```

Comment:
Is this check really needed? If an option hasn't been allocated, freeing it is no-op.

Reply:
Yes this check is needed. Using `free` on unallocated memory is in fact not no-op and will cause a segfault

Reply:
What unallocated memory?

Reply:
Oh, there is a single option: `p_vfile`

Reply:
It works if you use `empty_string_option` for that:
```diff
diff --git a/src/nvim/globals.h b/src/nvim/globals.h
index 45a31ce6b0..1cdb38f5f8 100644
--- a/src/nvim/globals.h
+++ b/src/nvim/globals.h
@@ -713,12 +713,6 @@ EXTERN char *escape_chars INIT( = " \t\\\"|");  // need backslash in cmd line
 
 EXTERN bool keep_help_flag INIT( = false);  // doing :ta from help file
 
-// When a string option is NULL (which only happens in out-of-memory situations), it is set to
-// empty_string_option, to avoid having to check for NULL everywhere.
-//
-// TODO(famiu): Remove this when refcounted strings are used for string options.
-EXTERN char *empty_string_option INIT( = "");
-
 EXTERN bool redir_off INIT( = false);        // no redirection for a moment
 EXTERN FILE *redir_fd INIT( = NULL);         // message redirection file
 EXTERN int redir_reg INIT( = 0);             // message redirection register
diff --git a/src/nvim/option.c b/src/nvim/option.c
index a0f58eaa16..72f8a5d2f3 100644
--- a/src/nvim/option.c
+++ b/src/nvim/option.c
@@ -3442,7 +3442,6 @@ static const char *did_set_option(OptIndex opt_idx, void *varp, OptVal old_value
   bool restore_chartab = false;
   bool value_changed = false;
   bool value_checked = false;
-  static bool opt_initialized[kOptIndexCount] = { 0 };
 
   optset_T did_set_cb_args = {
     .os_varp = varp,
@@ -3516,14 +3515,7 @@ static const char *did_set_option(OptIndex opt_idx, void *varp, OptVal old_value
     set_option_sctx(opt_idx, opt_flags, script_ctx);
   }
 
-  // Free options that are in allocated memory.
-  // NOTE: When initializing an option for the first time, don't try to free its value as it has
-  // not been allocated yet.
-  if (opt_initialized[opt_idx]) {
-    optval_free(old_value);
-  } else {
-    opt_initialized[opt_idx] = true;
-  }
+  optval_free(old_value);
 
   const bool scope_both = (opt_flags & (OPT_LOCAL | OPT_GLOBAL)) == 0;
 
diff --git a/src/nvim/option_vars.h b/src/nvim/option_vars.h
index bf72d4a18a..a4a7b8aaa8 100644
--- a/src/nvim/option_vars.h
+++ b/src/nvim/option_vars.h
@@ -350,6 +350,12 @@ enum {
 #define LISPWORD_VALUE \
   "defun,define,defmacro,set!,lambda,if,case,let,flet,let*,letrec,do,do*,define-syntax,let-syntax,letrec-syntax,destructuring-bind,defpackage,defparameter,defstruct,deftype,defvar,do-all-symbols,do-external-symbols,do-symbols,dolist,dotimes,ecase,etypecase,eval-when,labels,macrolet,multiple-value-bind,multiple-value-call,multiple-value-prog1,multiple-value-setq,prog1,progv,typecase,unless,unwind-protect,when,with-input-from-string,with-open-file,with-open-stream,with-output-to-string,with-package-iterator,define-condition,handler-bind,handler-case,restart-bind,restart-case,with-simple-restart,store-value,use-value,muffle-warning,abort,continue,with-slots,with-slots*,with-accessors,with-accessors*,defclass,defmethod,print-unreadable-object"
 
+// When a string option is NULL, it is set to empty_string_option,
+// to avoid having to check for NULL everywhere.
+//
+// TODO(famiu): Remove this when refcounted strings are used for string options.
+EXTERN char empty_string_option[] INIT( = "");
+
 // The following are actual variables for the options
 
 EXTERN char *p_ambw;             ///< 'ambiwidth'
@@ -772,7 +778,7 @@ EXTERN unsigned ve_flags;
 #define VE_NONEU       32U      // "NONE"
 EXTERN OptInt p_verbose;        ///< 'verbose'
 #ifdef IN_OPTION_C
-char *p_vfile = "";             ///< used before options are initialized
+char *p_vfile = empty_string_option;  ///< used before options are initialized
 #else
 extern char *p_vfile;           ///< 'verbosefile'
 #endif
```

Reply:
@zeertzjq  I wonder how future-proof this is, like whether this would work if we add Callback, Array and Dictionary types for options in the future

Reply:
surely those would be initialized to null?